File: ui/src/js/views/Projects/metadata.js
import { fetchMetadata as _fetchMetadata } from '../../metadata'

export function fetchMetadata(fetch, onSuccess, onError) {
  const data = {
    cookieCutters: null,
    environments: null,
    namespaces: null,
    projectLinkTypes: null,
    projectTypes: null,
    ready: false
  }
  let erred = false

  function onData(key, options) {
    data[key] = options
    if (Object.values(data).includes(null) === false && erred === false) {
      data.ready = true
      onSuccess(data)
    }
  }

  function onErr(value) {
    erred = true
    onError(value)
  }

  _fetchMetadata(
    fetch,
    '/cookie-cutters',
    false,
    null,
    null,
    (data) => {
      onData('cookieCutters', data)
    },
    onErr
  )

  _fetchMetadata(
    fetch,
    '/environments',
    true,
    'name',
    'name',
    (data) => {
      onData('environments', data)
    },
    onErr
  )

  _fetchMetadata(
    fetch,
    '/namespaces',
    true,
    'name',
    'id',
    (data) => {
      onData('namespaces', data)
    },
    onErr
  )

  _fetchMetadata(
    fetch,
    '/project-link-types',
    false,
    '',
    '',
    (data) => {
      onData('projectLinkTypes', data)
    },
    onErr
  )

  _fetchMetadata(
    fetch,
    '/project-types',
    true,
    'name',
    'id',
    (data) => {
      onData('projectTypes', data)
    },
    onErr
  )
}

File: legacy/imbi-api/imbi/endpoints/ui/metadata.py
import asyncio

from imbi.endpoints import base, cookie_cutters, environments, groups, \
    namespaces, fact_types, project_link_types, project_types


class RequestHandler(base.RequestHandler):

    NAME = 'metadata'

    async def get(self) -> None:
        """Return all metadata in a single request"""

        results = await asyncio.gather(
            self.postgres_execute(
                cookie_cutters.CollectionRequestHandler.COLLECTION_SQL,
                metric_name='cookie-cutters'),
            self.postgres_execute(
                environments.CollectionRequestHandler.COLLECTION_SQL,
                metric_name='environments'),
            self.postgres_execute(
                groups.CollectionRequestHandler.COLLECTION_SQL,
                metric_name='groups'),
            self.postgres_execute(
                namespaces.CollectionRequestHandler.COLLECTION_SQL,
                metric_name='namespaces'),
            self.postgres_execute(
                fact_types.CollectionRequestHandler.COLLECTION_SQL,
                metric_name='project-fact-types'),
            self.postgres_execute(
                project_link_types.CollectionRequestHandler.COLLECTION_SQL,
                metric_name='project-link-types'),
            self.postgres_execute(
                project_types.CollectionRequestHandler.COLLECTION_SQL,
                metric_name='project-types')
        )

        self.send_response({
            'cookie_cutters': results[0].rows,
            'environments': results[1].rows,
            'groups': results[2].rows,
            'namespaces': results[3].rows,
            'project_fact_types': results[4].rows,
            'project_link_types': results[5].rows,
            'project_types': results[6].rows
        })
